compiler/rustc_data_structures/src/functor.rs

```diff
@@ -55,6 +78,35 @@ impl<T> IdFunctor for Vec<T> {
         }
         self
     }
+
+    #[inline]
+    fn try_map_id<F, E>(mut self, mut f: F) -> Result<Self, E>
+    where
+        F: FnMut(Self::Inner) -> Result<Self::Inner, E>,
+    {
+        // FIXME: We don't really care about panics here and leak
+        // far more than we should, but that should be fine for now.
+        let len = self.len();
+        let mut error = Ok(());
+        unsafe {
+            self.set_len(0);
+            let start = self.as_mut_ptr();
+            for i in 0..len {
+                let p = start.add(i);
+                match f(ptr::read(p)) {
+                    Ok(value) => ptr::write(p, value),
+                    Err(err) => {
+                        error = Err(err);
+                        break;
+                    }
+                }
+            }
+            // Even if we encountered an error, set the len back
+            // so we don't leak memory.
+            self.set_len(len);
```

Comment:
I think this is wrong for the element that returned an Err. We "read" the value into `f`, but didn't "write" one back, so essentially that element was moved.

I'm not really well versed in unsafe code, so I think it would make sense to get a second opinion from someone else. cc @RalfJung @rust-lang/wg-unsafe-code-guidelines 

I am possibly also okay to land this PR with leaking memory and a FIXME and opened issue.

Reply:
Yes, I think you're right: this is UB.  Will give some thought how best way to handle this.